test/jobs/feed_crawler/downloader_test.rb

```diff
@@ -38,6 +38,40 @@ def test_should_persist_crawl_data_on_parse
       assert_equal 0, Parser.jobs.size, "should be empty because fingerprint will match"
     end
 
+    def test_should_persist_crawl_data_on_changed_conditional_headers_and_unchanged_fingerprint
+      new_etag = "new_etag"
+      new_last_modified = "new_last_modified"
+      download_fingerprint = "694b08e" # unchanged
+
+      existing_crawl_data = CrawlData.new({
+        etag: "old_etag",
+        last_modified: "old_last_modified",
+        checksum: download_fingerprint,
+        last_uncached_download: 1.hour.ago.to_i  # recent to not invoke ignore_http_caching?
+      })
+      refute existing_crawl_data.ignore_http_caching?, "should be false to test correctly"
+
+      stub_request(:get, /dhy5vgj5baket\.cloudfront\.net/).to_return(status: 404)
+      stub_request_file("atom.xml", @feed.feed_url,{
+        headers: {
+          "Etag" => new_etag,
+          "Last-Modified" => new_last_modified
+        }
+      })
+
+      Sidekiq::Testing.inline! do
+        Downloader.perform_async(@feed.id, @feed.feed_url, 10, existing_crawl_data.to_h)
+      end
+
+      assert_equal 0, Parser.jobs.size, "should be empty because fingerprint will match"
```

Comment:
One more subtle bug. Because of `Sidekiq::Testing.inline!`, `Parser.jobs.size` will always be zero because it runs immediately in the block. If you remove `Sidekiq::Testing.inline!`, and change line 63 to `Downloader.new.perform(@feed.id, @feed.feed_url, 10, existing_crawl_data.to_h)`, you can see that Parser does get queued up.

I think you'll also need to run `FeedCrawler::PersistCrawlData.new.perform`, in the test to actually persist the crawl_data. Right now it's only getting saved as a side effect of `Parser` running too.

Reply:
```patch
diff --git a/test/jobs/feed_crawler/downloader_test.rb b/test/jobs/feed_crawler/downloader_test.rb
index bf945df7..276242fb 100644
--- a/test/jobs/feed_crawler/downloader_test.rb
+++ b/test/jobs/feed_crawler/downloader_test.rb
@@ -59,9 +59,8 @@ module FeedCrawler
         }
       })
 
-      Sidekiq::Testing.inline! do
-        Downloader.perform_async(@feed.id, @feed.feed_url, 10, existing_crawl_data.to_h)
-      end
+      Downloader.new.perform(@feed.id, @feed.feed_url, 10, existing_crawl_data.to_h)
+      PersistCrawlData.new.perform
 
       assert_equal 0, Parser.jobs.size, "should be empty because fingerprint will match"
 
```

Reply:
Aha! Thanks! that's helpful. I think I see the problems here.

Reply:
Got it! I missed your suggestion but we arrived at the same place 😁